prov/verbs_ud/src/verbs_ud.h

```diff
@@ -0,0 +1,738 @@
+ * incoming message will start at an offset of 40 bytes
+ * into the buffer(s) in the scatter list.
+ */
+#define OFI_VERBS_UD_GRH_LENGTH 40
+
+struct ofi_verbs_ud_wr_entry {
+	struct ofi_verbs_ud_wr_entry_hdr	hdr;
+	char					grh_buf[];
+};
+
+static inline struct ofi_verbs_ud_wr_entry_hdr*
+ofi_verbs_ud_wr_entry_get(struct ofi_verbs_ud_buf_pool *pool)
+{
+	struct ofi_verbs_ud_wr_entry_hdr *buf;
+	struct fid_mr *mr = NULL;
+
+	buf = util_buf_alloc_ex(pool->pool, (void **)&mr);
+	if (unlikely(!buf))
+		return NULL;
+	buf->desc = fi_mr_desc(mr);
+	dlist_insert_tail(&buf->entry, &pool->buf_list);
+
+	return buf;
+}
+
+static inline void
+ofi_verbs_ud_wr_entry_release(struct ofi_verbs_ud_buf_pool *pool,
+			      struct ofi_verbs_ud_wr_entry_hdr *buf)
+{
+	dlist_remove(&buf->entry);
+	util_buf_release(pool->pool, buf);
+}
+
+#define OFI_VERBS_UD_WR_ENTRY_SIZE					\
+	(sizeof(struct ofi_verbs_ud_wr_entry) + OFI_VERBS_UD_GRH_LENGTH)
+
+static inline void
+ofi_verbs_ud_mr_buf_close(void *pool_ctx, void *context)
+{
+	/* We would get a (fid_mr *) in context but
+	 * it is safe to cast it into (fid *) */
+	fi_close((struct fid *)context);
+}
+
+static inline int
+ofi_verbs_ud_mr_buf_reg(void *pool_ctx, void *addr,
+			size_t len, void **context)
+{
+	int ret;
+	struct fid_mr *mr;
+	struct fid_domain *domain = (struct fid_domain *)pool_ctx;
+
+	ret = fi_mr_reg(domain, addr, len,
+			FI_SEND | FI_RECV | FI_READ | FI_WRITE,
+			0, 0, 0, &mr, NULL);
+	*context = mr;
+	return ret;
+}
+
+static inline void
+ofi_verbs_ud_pool_wr_entry_cancel(struct ofi_verbs_ud_buf_pool *pool)
+{
+	struct dlist_entry *entry;
+	struct ofi_verbs_ud_wr_entry_hdr *buf;
+	entry = pool->buf_list.next;
+	buf = container_of(entry, struct ofi_verbs_ud_wr_entry_hdr,
+			   entry);
+	ofi_verbs_ud_wr_entry_release(pool, buf);
+	
+}
+
+static inline void
+ofi_verbs_ud_pool_destroy(struct ofi_verbs_ud_buf_pool *pool)
+{
+	if (pool->cancel_hndlr) {
+		while (!dlist_empty(&pool->buf_list))
+			pool->cancel_hndlr(pool);
+	}
+
+	util_buf_pool_destroy(pool->pool);
+}
+
+static inline int
+ofi_verbs_ud_pool_create(struct ofi_verbs_ud_pool_attr *attr,
+			 struct ofi_verbs_ud_buf_pool *pool)
+{
+	pool->pool = util_buf_pool_create_ex(
+			attr->size, 16, 0, attr->count,
+			attr->alloc_hndlr, attr->free_hndlr,
+			attr->pool_ctx
+	);
+	if (!pool->pool) {
+		VERBS_UD_WARN(FI_LOG_EP_DATA,
+			      "Unable to create buf pool\n");
+		return -FI_ENOMEM;
+	}
+	pool->cancel_hndlr = attr->cancel_hndlr;
+	dlist_init(&pool->buf_list);
+
+	return FI_SUCCESS;
+}
+
+/* ~Verbs UD Pool functionality */
+
+#define OFI_VERBS_UD_DEFAULT_NS_PORT	5678
+
+extern struct gl_data {
+	int	fork_unsafe;
+	struct {
+		char	*device_name;
+		int	port_number;
+	} device;
+	int	def_inject_size;
+	int	def_tx_ctx_size;
+    	int	def_rx_ctx_size;
+	int	def_tx_iov_limit;
+	int	def_rx_iov_limit;
+	char	*interface_name;
+	int	use_ipoib_dev_only;
+	int	use_name_server;
+	int	name_server_port;
+	int	separate_send_recv_cq;
```

Comment:
This doesn't seem necessary and should fall out by how the app assigns cq's to endpoints.